motionplan/collision.go

```diff
@@ -311,3 +313,116 @@ func createUniqueCollisionMap(geoms []spatial.Geometry) (map[string]spatial.Geom
 	}
 	return geomMap, nil
 }
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan.
+func CheckPlan(
+	frame referenceframe.Frame,
+	plan [][]referenceframe.Input,
+	obstacles []*referenceframe.GeometriesInFrame,
+	fs referenceframe.FrameSystem,
+) (bool, error) {
+	// ensure that we can actually perform the check
+	if len(frame.DoF()) != len(plan[0]) {
+		return false, errors.New("frame DOF length must match inputs length")
+	}
+	// check if we are working with ptgs
+	ptgProv, ok := frame.(tpspace.PTGProvider)
+	if ok {
+		return checkPtgPlan(frame, plan, obstacles, fs, ptgProv.PTGs())
+	}
+
+	if len(plan) < 2 {
```

Comment:
It feels weird to me to have two conventions.  @biotinker is this something that will change with your new version of tp space?

Reply:
The current/future implementation of tp-space book-ends the plan with a zeroed out configuration, so the minimum plan length that would be output by tp-space is 3.

However, that's a coincidence of implementation, and a single-element tp-space plan is indeed valid and we should be able to check it for issues.

Reply:
We're going to need to standarize the way these plans are defined as a part of the scope to expose plans to the user.  My preference would be to keep the initial state in the plan such that the minimum length is always 2

Reply:
I am ok with either.
If the minimum length remains as 2 we can continue doing the following
```
opt.CheckSegmentAndStateValidity(
			&Segment{
				StartConfiguration: now,
				EndConfiguration:   nextStep,
				Frame:              sf,
			},
			opt.Resolution,
```

it is also fine to have the minimum length be 1 since we can always do `referenceframe.startingPositions(fs)`.

Reply:
[q]: would having minimum length = 1 make it more complicated to expose motion plans to the user?

Reply:
I take it back.
I don't think we can rely on `referenceframe.startingPositions(fs)` since that will return a zero'd map.
As an example, for arms we cannot assume that they will start in the home pose when we perform.
instead we would need to use CurrentInputs from the framesystem service

Reply:
> I'm inclined to think we should only pass in the plan left to execute to this function, and make no guarantees about the nature of the first node in the path. Otherwise, we're checking lots of things we've already passed and are irrelevant, which is unnecessary.

I agree with you on this, and also that currentInputs will need to be passed in as an additional argument to fulfill this functionality.

> we need to adjust the positions that we check based on the actual locations of things, not the predicted ones from the plans.

I want to clarify what you mean here.
When you say "adjust positions ... based on the actual locations of things" do you mean that we should have another argument added to the function which represents the current `errorState` value as a `pose` of the robot?

Reply:
> When you say "adjust positions ... based on the actual locations of things" do you mean that we should have another argument added to the function which represents the current errorState value as a pose of the robot?

Exactly.  RSDK-4502 will implement the functionality that will provide such a pose; this function should be a consumer of that pose.

Reply:
ok this stgm - yay 😄 I am happy that things are coming together.

I will update [RSDK-4387](https://viam.atlassian.net/browse/RSDK-4387) to say that it is blocked by RSDK-4502.

I want to make sure we are aligned though.. @raybjork what do you think about 
>I'm inclined to think we should only pass in the plan left to execute to this function, and make no guarantees about the nature of the first node in the path. Otherwise, we're checking lots of things we've already passed and are irrelevant, which is unnecessary.

I know that you previously commented that
> My preference would be to keep the initial state in the plan such that the minimum length is always 2

Reply:
4387 is not blocked by 4502.

This function will simply take a pose. `ErrorState` does not need to exist yet for that functionality to be written.

The thing in the builtin motion service which would be calling both `ErrorState` and `CheckPlan` is blocked on both those tickets, but neither `CheckPlan` nor `ErrorState` should block one another.

Agree we should align with @raybjork regarding the plan passed in.

Reply:
Sorry to have missed this earlier.  I think it makes sense to pass in only what remains since that is the relevant part.  However, it shouldn't affect the implementation of this helper function because the subset of a plan is itself also a plan

Reply:
I'm not sure we want a min path length of 1 rather than 2.  Having the start input present in the path means we can treat a plan as standalone and not need to make other requests to figure out where it originates from.  For now bifurcating this way I think is fine until we figure out how we're going to do this long term